[Apps] Prototype backend app scaffolding

snoe opened the pull request:
Given a set of table-ids, we try to build out an app. Ideally the produced scaffold is done on the frontend and passed in, in order to:

1. avoid code duplication
2. allow the front end to maintain ownership of visualization_settings
3. avoid regressions on frontend code changes

However, this currently builds the scaffold on the backend due to FE dev bandwidth. In theory, it is done in such a way that it both matches the FE as closely as possible as well as becomes easy to change this code to accept a scaffold rather than generating one itself.

It uses `scaffold-target` to map cards and dashboards in the scaffold with the inserted ids.

snoe committed:
[Apps] Prototype backend app scaffolding

WIP

Given a set of table-ids, we try to build out an app. Ideally the
produced scaffold is done on the frontend and passed in to

1. avoid code duplication
2. allow the front end to maintain ownership of visualization_settings
3. avoid regressions on frontend code changes

However, this currently builds the scaffold on the backend due to FE dev
bandwidth. In theory, it is done in such a way that it both matches the
FE as closely as possible as well as becomes easy to change this code to
accept a scaffold rather than generating one itself.

It uses `scaffold-target` to map cards and dashboards in the scaffold
with the inserted ids.

Things that are still being worked out elsewhere:
1. The shape of nav-items
2. The shape of implicit action buttons

snoe committed:
Sort namespace

metamben commented on src/metabase/api/app.clj:
```diff
@@ -68,4 +76,139 @@
+
+(defn- generate-scaffold
+  [app-name table-ids]
+  (let [tables (hydrate (db/select Table :id [:in table-ids]) :fields)
```
- Should we check if all `table-ids` actually point to a table?
- Maybe it would be worth converting `table-ids` into an ordered set, just to make sure we don't generate duplicates. A normal set could be good enough too, we just have to make sure that the first element get's the special treatment it needs.

metamben commented on src/metabase/api/app.clj:
```diff
@@ -68,4 +76,139 @@
+  [app-name table-ids]
+  (let [tables (hydrate (db/select Table :id [:in table-ids]) :fields)
+        table-id->table (m/index-by :id tables)
+        page-type-display {"list" {:name "List"
+                                   :display "list"}
+                           "detail" {:name "Detail"
+                                     :display "object"}}]
```
Should we worry about i18n here?

snoe replied:
good catch.

metamben commented on src/metabase/api/app.clj:
```diff
@@ -68,4 +76,139 @@
+          ;; We create the dashboards (without dashcards) so we can replace scaffold-target-id elsewhere
+          scaffold-target->id (reduce
+                                (fn [accum {:keys [scaffold-target] :as page}]
+                                  (when-not scaffold-target
+                                    (throw (ex-info (i18n/tru "A scaffold-target was not provided for Page: {0}" (:name page))
+                                                    {:status-code 400})))
```
Why do we check pages but not cards?

metamben commented on src/metabase/api/app.clj:
```diff
@@ -68,4 +76,139 @@
+  [structure scaffold-target->target-id]
+  (walk/postwalk
+    (fn [node]
+      (if (and (vector? node) (= "scaffold-target-id" (first node)))
```
I suppose this check is not really necessary, the next line will not match anyway if this does not. If you keep this check, you can make the `scaffold-target->target-id` map simpler by omitting the tag around the id.

snoe committed:
Update and fix scaffold based on demo

snoe commented on src/metabase/api/app.clj:
```diff
@@ -68,4 +76,139 @@
+                               :query {:source_table table-id}}})
+     :pages (for [table-id table-ids
+                  :let [table (get table-id->table table-id)
+                        pk-field-id (first (filter (comp #(= :type/PK %) :semantic_type) (:fields table)))]
```
correct. I can add in handling for > 1

snoe committed:
Address review comments

Deduplicate table-ids and make sure they are valid.

i18n page suffixes.

Check that tables have exactly one primary key column.

Check that card scaffold has a `scaffold-target`

Remove redudant check in scaffold-target replacement that the map lookup took care of.

snoe committed:
Use Integer rather than Long for h2

metamben commented on src/metabase/api/app.clj:
```diff
@@ -68,4 +76,151 @@
+        tables (hydrate (db/select Table :id [:in table-ids]) :fields)
+        _ (when (not= (count table-ids) (count tables))
+            (throw (ex-info (i18n/tru "Some tables could not be found. Given: {0} Found: {1}"
+                                      (pr-str table-ids)
+                                      (pr-str (map :id tables)))
```
It would be fairly easy to output just the dangling ones.
